syntax = "proto3";
import "google/protobuf/any.proto";

option java_package = "se.iimetra.dataflowgram.worker";

message Update {}
message Updated {}

message ExecutionRequest {
    string function = 1;
    repeated string args = 2;
    map<string, string> params = 3;
}

message ExecutionResult {
    oneof value {
        string msg = 1;
        string ref = 2;
    }
}

message OutRequest {
    string ref = 1;
    string type = 2;
}

message OutResult {
    string value = 1;
}

message InRequest {
    string value = 1;
    string type = 2;
}

message InResult {
    string ref = 1;
}

service Executor {
    rpc Execute (ExecutionRequest) returns (stream ExecutionResult) {}
    rpc OutData (OutRequest) returns (OutResult) {}
    rpc InData (InRequest) returns (InResult) {}
    rpc UpdateLib (Update) returns (Updated) {}
}
